### DISCUSSION TITLE
Common prefix resolution

### DISCUSSION THREAD

**Discussion by luan-xiaokun** [antlr/antlr4#3939] [Q&A] [answered]
Hi guys, I'm new to ANTLR4, I'm working on a project where the language supports the C++ style template, i.e. both type alias and expression can be passed to the template instantiation.

The current parse rule looks like this:
```
templatedInstance:
    scopedID ('<' typeOrExpr (',' typeOrExpr )* '>')? ;
typeOrExpr:
    type | expr ;
type:
    scopedID | ... | ... ;
expr:
    ID | ... | ... ;
scopedID:
    (scopes+=ID '.')* identifier=ID ;
```
Clearly `ID` is a common prefix of `type` and `expr`, so these rules seem to be ambitious, and I was expecting that ANTLR would raise errors. But instead, nothing goes wrong.

I'm wondering how ANTLR would recognize an `ID` in such case, or am I missing something here? Any advice on how to implement this in ANTLR is also welcome. Thanks in advance.

**Comment by kaby76** [comment, marked as answer]
Antlr will take almost any grammar and generate a parser for it. That parser will handle the common prefix. However, the grammar is ambiguous, so Antlr will choose one derivation or the other (likely through "type"). You should fix the grammar.

**Reply by luan-xiaokun** [reply]
Thanks for your reply! Is there any common approach to support such template design? One way I could come up with is to define a new group of grammar rules to capture whatever looks like a type or an expression, but this may require more elaborated validation. Is there some other way to do this? Any link or reference about this would be very helpful.

**Reply by kaby76** [reply]
There is no tool that does this automatically, but I am working towards one. The procedure is easy for some situations:
* Unfold the "called" rules into the RHS of the "calling" rule. E.g., substitute the RHS of "type" to where it was used in "typeOrExpr", etc.
* Apply [grouping](https://github.com/kaby76/Domemtech.Trash/blob/main/doc/refactoring.md#group-alts) to the rule containing all "ID"s.

Note, this will reorder the alts, and thus may change the parse significantly.

But, apply this to your grammar might be more challenging.

**Reply by luan-xiaokun** [reply]
Many thanks to your detailed reply,  I will try it out.
